### DISCUSSION TITLE
Chapter 07 exercise template notebook

### DISCUSSION THREAD

**Discussion by wlwlaa** [mrdbourke/pytorch-deep-learning#920] [Q&A] [answered]
* 07_pytorch_experiment_tracking_exercise_template.ipynb

## My error

In exercise 1 I got some problems with CUDA device memory in google colab during training a few different models

## My code
```python
%time
# Отследим номер эксперимента
experiment_number = 0

# Пройдем по всем моделям, количеству эпох и даталоадерам из множества
for model_name in model_list:
  for dataloader_name, dataloader in train_dataloader_dict.items():
    for epochs in epochs_list:

      # Обновим информацию
      experiment_number += 1
      print(f'[INFO] {{{experiment_number}}} Тестируем {model_name} на {epochs} эпох на {dataloader_name} данных...')

      # Инициализируем модель
      match model_name:
        case 'effnetb3':
          model = createEffNetB3()
        case 'effnetb7':
          model = createEffNetB7()
        case 'effnetv2l':
          model = createEffNetV2_L()

      # Настроим loss func & optimizer
      loss_fn = nn.CrossEntropyLoss()
      optimizer = torch.optim.Adam(
          model.parameters(),
          lr=0.0005
      )

      # Обучим модель с выбранными параметрами
      train(
          model,
          dataloader,
          test_dataloader,
          optimizer,
          loss_fn,
          epochs,
          device,
          create_writer(
              dataloader_name,
              model_name,
              f'{epochs}_epochs'
          )
      )

      # Сохраним обученную модель
      utils.save_model(
          model,
          'models/',
          f'{model_name}_{dataloader_name}_{epochs}_epochs.pth'
      )

      # Очистим кэш видеоустройства
      if device == 'cuda':
        torch.cuda.empty_cache()

      print('-' * 100, '\n')
``` 

Output: 
OutOfMemoryError: CUDA out of memory. Tried to allocate 20.00 MiB. GPU 0 has a total capacity of 14.75 GiB of which 3.06 MiB is free. Process 136418 has 14.74 GiB memory in use. Of the allocated memory 14.28 GiB is allocated by PyTorch, and 345.42 MiB is reserved by PyTorch but unallocated. If reserved but unallocated memory is large try setting PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True to avoid fragmentation.  See documentation for Memory Management  (https://pytorch.org/docs/stable/notes/cuda.html#environment-variables)

## What I've tried so far

I added `torch.cuda.empty_cache()` at the end of each iteration but this helped partially

**Comment by IlanVinograd** [comment]
я думаю что у вашей видеокарте не хватает памяти для выполнения операции поэтому указывает на CUDA out of memory.

попробуй уменьшить размера пакета (batch size).

или как ты и сказал с помощью `torch.cuda.empty_cache()` после каждой эпохи обучения.
она освобождает память тем самым не должно быть переполнение.

скажи если помогло 😄

**Reply by wlwlaa** [reply]
Я добавил в конец каждой итерации
```Python
del model
if device == 'cuda':
  torch.cuda.empty_cache()
  gc.collect()
```
Также настройку окружения `os.environ['PYTORCH_CUDA_ALLOC_CONF'] = 'expandable_segments:True'` 
Но все это не помогло, что-то продолжает забивать память CUDA

**Reply by IlanVinograd** [reply]
скинь полный код

**Reply by wlwlaa** [reply]
https://colab.research.google.com/drive/1jjOUUBgoGdgfJOZHogAA7XNYimkFxWFD?usp=sharing

**Reply by IlanVinograd** [reply, marked as answer]
![image](https://github.com/mrdbourke/pytorch-deep-learning/assets/128734827/8e325907-11d5-4494-a2f9-bb5b83e61aec)
BATCH_SIZE = 16
попробуй

**Reply by IlanVinograd** [reply]
ты в сети?

**Reply by wlwlaa** [reply]
Да. Уменьшение батч сайза пока помогло, вторая модель в процессе обучения, ошибок нет

**Reply by IlanVinograd** [reply]
я у себя попробовал уменьшил батчи на 16 
![image](https://github.com/mrdbourke/pytorch-deep-learning/assets/128734827/db0f4c2d-57ef-4f9d-aea4-acbf5b73b2e5)
видишь сколько памяти забивает +- 2/3 которой есть.

**Reply by wlwlaa** [reply]
<img width="517" alt="Снимок экрана 2024-05-08 в 00 36 08" src="https://github.com/mrdbourke/pytorch-deep-learning/assets/112896514/dd09f99a-ad21-44c0-8bdd-cd8bf8133cb3">


Аналогично.

**Reply by IlanVinograd** [reply]
но смотри в момент когда 21 модель начинает  она сразу выше 15GB тут нужно или круче видюхи с больше памяти или уменьшать и больше ждать.
я ждал 45 минут пока она тестирует 😆

**Reply by IlanVinograd** [reply]
я снова проверил когда вот эта модель > effnetv2l она выдает ошибку памяти это очень тяжёлая  модель или упрости её  саму типа её  слои или поменяй на другую а так остальные модели проходят.
![image](https://github.com/mrdbourke/pytorch-deep-learning/assets/128734827/bcc9350a-f272-4937-ae32-84e69414ec47)

**Reply by IlanVinograd** [reply]
![Screenshot 2024-05-08 004511](https://github.com/mrdbourke/pytorch-deep-learning/assets/128734827/b74cdaef-6992-4e1e-b00c-aaa4e659b090)

**Reply by wlwlaa** [reply]
<img width="976" alt="image" src="https://github.com/mrdbourke/pytorch-deep-learning/assets/112896514/d796094b-a723-4bf9-9ea5-6334c230a763">

Только закончилось обучение. Заняло пол часа и EffNet V2 L все же с батч сайзом 16 вытянуло

**Reply by IlanVinograd** [reply]
Честно говоря я сам в шоке, у меня не вытянула.
Вот поставлю снова на ночь может получится.

**Reply by wlwlaa** [reply]
ты покупал вычислительные единицы или пользуешься бесплатной версией?

**Reply by IlanVinograd** [reply]
Бесплатной

**Reply by wlwlaa** [reply]
у тебя часто заканчиваются вычислительные единицы на GPU? у меня в последнее время довольно часто прекращается доступ к видеоускорителю из-за этого и сейчас волновался что не успею обучить

**Reply by IlanVinograd** [reply]
Ну я сам вообще не пользуюсь colab, просто использую сам процессор/видеокарту на моём ПК.

так что не замечал.
